src/week2.hs

```diff
@@ -0,0 +1,176 @@
+isConsistent :: Move -> Code -> Bool
+
+isConsistent (Move c x y) c' 
+    |(exactMatches c c') == x && ((matches c c') - (exactMatches c c')) == y = True
+    |otherwise                                                               = False
+    
+-- Or:
+isConsistent1 :: Move -> Code -> Bool
+isConsistent1 (Move c x y) c'  = e == x && (m - e) == y
+    where e = exactMatches c c'
+          m = matches c c'
+
+-- Exercise 5 -----------------------------------------
+-- Write a function that filters a list of Codes to only contain those that are consistent with a given Move
+filterCodes :: Move -> [Code] -> [Code]
+filterCodes _ [] = []
+filterCodes m (c:cs) = if (isConsistent m c) then (c : filterCodes m cs)
+                                             else filterCodes m cs
+-- or, more concisely:
+filterCodes1 :: Move -> [Code] -> [Code]
+filterCodes1 m c = filter (isConsistent m) c
+
+-- Exercise 6 -----------------------------------------
+-- Write a function that outputs all possible code permutations of a given length
+
+pegs :: Code
+pegs = [Red, Green, Blue, Yellow, Orange, Purple]
+baseCodes :: [Code]
+baseCodes = [[]]
+
+-- This function takes in a Code of length k and a list of j colors and outputs
+-- a list of j codes where each code has length k+1 and originates from the input code plus one color from the color list
+addToCode :: Code -> Code -> [Code]
+addToCode [] _ = []
+addToCode (p:ps) c = [(p:c)] ++ (addToCode ps c)
+
+-- This function takes in a LIST of codes and does the operation described above for each code
+addToListCode :: Code -> [Code] -> [Code]
+addToListCode p c = concatMap (addToCode p) c
+
+-- Main function: outputs a list of all posible codes of a given length
+allCodes :: Int -> [Code]
+allCodes 0 = baseCodes
+allCodes n = addToListCode pegs (allCodes (n-1))
+
+-- Exercise 7 -----------------------------------------
+
+-- Write a MasterMind solver
+solve :: Code -> [Move]
+
+possibilities :: [Code]
+possibilities = allCodes 4
+
+-- helper function to reverse list (so that our Move list goes least recent -> most recent guess)
+reverseListHelper :: [a] -> [a] -> [a]
+reverseListHelper [] acc = acc
+reverseListHelper (a:as) acc = reverseListHelper as (a:acc)
+
+reverseList :: [a] -> [a]
+reverseList x = reverseListHelper x []
+
+-- another helper
+-- takes in a password, a list of possible guesses, and a list of moves (initially empty), 
+-- and goes through the possible guesses, filtering out inconsistent ones, until the password is found
+-- Output is a list of the moves made until we guess correctly
+solveHelper :: Code -> [Code] -> [Move] -> [Move]
+solveHelper _ [] m = m
+solveHelper c (p:ps) m
+    |exactMatches p c == 4 = reverseList((getMove c p) : m)
+    |otherwise             = solveHelper c (filterCodes (getMove c p) ps) ((getMove c p) : m)
+
+
+-- main function uses solveHelper starting with all possible guesses of size 4 and an empty list of Moves
+solve c = solveHelper c possibilities []
```

Comment:
this isn't a strict requirement but I think it's preferable to have this definition of `solve` right below its type signature. it's more a matter of taste if you want to move this up above or move the type signature here.